rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,427 @@
+pub const RECORD_VERSION_NUMBER: &str = "recordVersionNumber";
+pub const IS_RELEASED: &str = "isReleased";
+pub const LEASE_DURATION: &str = "leaseDuration";
+pub const DATA: &str = "data";
+
+mod expressions {
+    pub const ACQUIRE_LOCK_THAT_DOESNT_EXIST: &str = "attribute_not_exists(#pk)";
+
+    pub const PK_EXISTS_AND_IS_RELEASED: &str = "attribute_exists(#pk) AND #ir = :ir";
+
+    pub const PK_EXISTS_AND_RVN_MATCHES: &str = "attribute_exists(#pk) AND #rvn = :rvn";
+
+    pub const PK_EXISTS_AND_OWNER_RVN_MATCHES: &str =
+        "attribute_exists(#pk) AND #rvn = :rvn AND #on = :on";
+
+    pub const UPDATE_IS_RELEASED_AND_DATA: &str = "SET #ir = :ir, #d = :d";
+
+    pub const UPDATE_IS_RELEASED: &str = "SET #ir = :ir";
+}
+
+mod vars {
+    pub const PK_PATH: &str = "#pk";
+    pub const RVN_PATH: &str = "#rvn";
+    pub const RVN_VALUE: &str = ":rvn";
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+    pub const OWNER_NAME_PATH: &str = "#on";
+    pub const OWNER_NAME_VALUE: &str = ":on";
+    pub const DATA_PATH: &str = "#d";
+    pub const DATA_VALUE: &str = ":d";
+}
+
+mod options {
+    pub const PARTITION_KEY_VALUE: &str = "DYNAMO_LOCK_PARTITION_KEY_VALUE";
+    pub const TABLE_NAME: &str = "DYNAMO_LOCK_TABLE_NAME";
+    pub const OWNER_NAME: &str = "DYNAMO_LOCK_OWNER_NAME";
+    pub const LEASE_DURATION: &str = "DYNAMO_LOCK_LEASE_DURATION";
+    pub const SLEEP_MILLIS: &str = "DYNAMO_LOCK_SLEEP_MILLIS";
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        fn str_env(key: &str, default: String) -> String {
+            std::env::var(key).unwrap_or(default)
+        }
+
+        fn u64_env(key: &str, default: u64) -> u64 {
+            std::env::var(key)
+                .ok()
+                .and_then(|e| e.parse::<u64>().ok())
+                .unwrap_or(default)
+        }
+
+        Self {
+            partition_key_value: str_env(options::PARTITION_KEY_VALUE, "delta-rs".to_string()),
+            table_name: str_env(options::TABLE_NAME, "delta_rs_lock_table".to_string()),
+            owner_name: str_env(options::OWNER_NAME, Uuid::new_v4().to_string()),
+            lease_duration: u64_env(options::LEASE_DURATION, 20),
+            sleep_millis: u64_env(options::SLEEP_MILLIS, 1000),
+        }
+    }
+}
+
+impl DynamoLock {
+    pub fn new(client: DynamoDbClient, opts: Options) -> Self {
+        Self { client, opts }
+    }
+
+    pub async fn acquire_lock(&self) -> Result<LockItem, DynamoError> {
+        let mut state = AcquireLockState {
+            client: self,
+            active_lock: None,
+            started: Instant::now(),
+            timeout_in: Duration::from_millis(self.opts.sleep_millis),
+        };
+
+        loop {
+            match state.try_acquire_lock().await {
+                Ok(lock) => return Ok(lock),
+                Err(DynamoError::ConditionalCheckFailed) => {
+                    if state.has_timed_out() {
```

Comment:
i am curious when will we ever hit this branch, it looks like if the lock is not expired, we won't reach timeout because lock lookup time is set before we extend timeout_in. when the lock is expired, we will get the new lock and refresh timeout_in with the release duration.

Reply:
that's why we need a cached/active lock, if rvn of retrieved existing lock matches the one we cached then we simply try again later, hence lookup_time will remain the same and eventually we'll get `has_timed_out()` to be true. Otherwise, if rvn of retrieved lock changed, then either the existing lock has been refreshed (we don't have a function for it now) or the owner has died

Reply:
Since we set the lock lookup time slightly earlier before extending the timeout, wouldn't it hit `cached.is_expired()` branch first before we return to do the timeout check? This branch calls `upsert_expired_lock`, which extends the timeout again before returning to the try loop. Wouldn't this make the `state.has_timed_out` check pass?

After thinking more on this, it looks like we could hit the timeout branch in the following scenario:
1. try_acquire_lock returns an active lock
2. cached lock lookup time set to now
3. timeout_in extended by lease time
4. lock expired
5. conditional lock upsert failed due to another concurrent writer got the lock first
6. back to try loop with `ConditionalCheckFailed` raised
7. timeout check failed, return timeout error

In this scenario, we exit the loop after trying to acquire the lock once. Is this an expected behavior?
In your doc string, you added `Attempts to acquire a lock until it either acquires the lock or the error is found.`, do we consider concurrent lock acquisition failures to be a failure that would break the loop?

Reply:
Okay, so I get deterministic TimedOut for one of the concurrent calls (imagine 2 workers try to acquire already expired lock) when using unit tests. However when I tested it manually everything worked. This is because in unit tests they're started in parallel however when running manually there's time between those calls, and if that time is more than a `sleep_millis` (1 sec) then everything is okay, however is less (unit tests) it fails with `TimedOut`. 

That is because 2nd, the failed worker already reaches its timeout before one more call to the dynamodb to find out that the lock has been changed. 

I think I should stop reinventing a wheel here and proceed with original impl, since they return `None` in the situation  above, and leave to the caller decide on what to do next, either fail or continuing  on trying. Also, if we'll go with `Option<LockItem>` as a result of `acquire_lock` we'll get 1-to-1 match with original version which which will allow as to move dynamodb_lock into separate library.

Reply:
@houqp
>Is this an expected behavior?

Also very good catch on this since I was sure otherwise until wrote a unit tests

Reply:
@houqp I've updated are updated and are the same as in origin's since both impl are the same now